src/gedi_subset/subset.py

```diff
@@ -424,49 +426,50 @@ def main(
     maap = MAAP("api.maap-project.org")
     cmr_host = "cmr.earthdata.nasa.gov"
 
-    IOResult.do(
-        subsets
-        for aoi_gdf in impure_safe(gpd.read_file)(aoi)
-        # Use wildcards around DOI value because some collections have incorrect
-        # DOI values. For example, the L2B collection has the full DOI URL as
-        # the DOI value (i.e., https://doi.org/<DOI> rather than just <DOI>).
-        for collection in find_gedi_collection(
-            maap, dict(cmr_host=cmr_host, doi=f"*{doi}*", cloud_hosted="true")
-        )
-        for granules in impure_safe(maap.searchGranule)(
-            cmr_host=cmr_host,
-            collection_concept_id=collection["concept-id"],
-            bounding_box=",".join(fp.map(str)(aoi_gdf.total_bounds)),  # pyright: ignore
-            limit=limit,
-            **(dict(temporal=temporal) if temporal else {}),
-        )
-        for subsets in subset_granules(
-            maap,
-            aoi_gdf,
-            lat,
-            lon,
-            beams,
-            [c.strip() for c in columns.split(",")],
-            query,
-            output_dir,
-            dest,
-            (logging_level,),
-            fp.filter(granule_intersects(aoi_gdf.unary_union))(granules),
-            fsspec_kwargs,
-            processes,
-        )
-    ).bind_ioresult(
-        lambda subsets: (
-            IOSuccess(subsets)
-            if subsets
-            else IOFailure(ValueError(f"No granules intersect the AOI: {aoi}"))
+    gpq_paths = unsafe_perform_io(
+        IOResult.do(
+            subsets
+            for aoi_gdf in impure_safe(gpd.read_file)(aoi)
+            # Use wildcards around DOI value because some collections have incorrect
+            # DOI values. For example, the L2B collection has the full DOI URL as
+            # the DOI value (i.e., https://doi.org/<DOI> rather than just <DOI>).
+            for collection in find_gedi_collection(
+                maap, dict(cmr_host=cmr_host, doi=f"*{doi}*", cloud_hosted="true")
+            )
+            for granules in impure_safe(maap.searchGranule)(
+                cmr_host=cmr_host,
+                collection_concept_id=collection["concept-id"],
+                bounding_box=",".join(
+                    fp.map(str)(aoi_gdf.total_bounds)
+                ),  # pyright: ignore
+                limit=limit,
+                **(dict(temporal=temporal) if temporal else {}),
+            )
+            for subsets in subset_granules(
+                maap,
+                aoi_gdf,
+                lat,
+                lon,
+                beams,
+                [c.strip() for c in columns.split(",")],
+                query,
+                output_dir,
+                dest,
+                (logging_level,),
+                fp.filter(partial(granule_intersects, aoi_gdf.unary_union))(granules),
+                fsspec_kwargs,
+                processes,
+            )
         )
```

Comment:
Do we want to keep f"No granules intersect the AOI: {aoi}" or are we now saying that there are multiple reasons why a subset can be empty?

Reply:
There are at least 2 reasons the result can be empty: (1) no intersection with the AOI, or (2) nothing satisfies the query. The previous "No granules intersect the AOI" error message accounted for only reason 1, not reason 2, even though that message would be used in both cases.

We could add a little bit more logic to distinguish between the 2 cases, and print the appropriate message. That could be helpful for a user to debug why they're getting an empty result because they would know whether the problem is the AOI + temporal (we could enhance the message to say no intersection with the AOI _within the temporal range_) or the query (i.e., granules intersect with AOI + temporal, but no data within the granules satisfy the query).

What do you think? Should I add a bit of extra logic for that?

Reply:
Ah, thought I had responded. I think that anything that might be useful to the users might save you a few minutes down the road when they encounter empty subsets.